etl/topline_summary.kp/knowledge.md

```diff
@@ -0,0 +1,447 @@
+        os_udf('os').alias('os'),
+        search_udf('search_counts').alias('search_counts'),
+        hour_udf(F.col('subsession_length').cast('double')).alias('hours')
+    ]
+
+    report_df = (
+        main_summary_df
+            .filter(start_date <= F.col('submission_date_s3'))
+            .filter(F.col('submission_date_s3') <= end_date)
+            .filter(F.col('app_name') == 'Firefox')
+            .select(columns)
+    )
+    
+    return report_df
+```
+### telemetry-pings where docType = 'crash'
+We are missing the crash pings in the main summary, but we can filter the raw pings manually.
+
+
+```python
+from moztelemetry import get_pings_properties
+from moztelemetry.dataset import Dataset
+
+
+def get_crash_df(start_date, end_date):
+    """ Return a dataframe containing relevent columns from crash pings. """t
+    ping_rdd = (
+        Dataset
+            .from_source("telemetry")
+            .where(docType='crash')
+            .where(appName='Firefox')
+            .where(submissionDate=lambda x: start_date <= x <= end_date)
+            .records(sc)
+    )
+
+    crash_rdd = get_pings_properties(ping_rdd, {
+            "country": "meta/geoCountry", 
+            "channel": "meta/appUpdateChannel",
+            "os": "meta/os"
+        })
+
+    crash_schema = StructType([
+            StructField("country", StringType(), True),
+            StructField("os", StringType(), True),t
+            StructField("channel", StringType(), True)
+        ])
+
+    # this guarantees the order in the crash_schema
+    def to_tuple(row):
+        return row['country'], row['os'], row['channel']
+
+    return sqlContext.createDataFrame(crash_rdd.map(to_tuple), crash_schema)
+```
+    Unable to parse whitelist (/home/hadoop/anaconda2/lib/python2.7/site-packages/moztelemetry/histogram-whitelists.json). Assuming all histograms are acceptable.
+
+
+## Queries
+
+
+```python
+def get_easy_aggregates(report_df, crash_df, groups, report_date):
+    """ Return the aggregates of the number of hours, crashes, and search counts. """
+    
+    # Clean the columns of the crash aggregate to match the easy ones
+    crash_agg_df = (
+        crash_df
+            .select(
+                country_udf('country').alias('country'),
+                channel_udf('channel').alias('channel'), 
+                os_udf('os').alias('os'))
+            .groupby(groups) 
+            .agg(F.count('*').alias('crashes'))
+    )
+
+    # Join the hours and search aggregates with the crash aggregates
+    easy_aggregate_df = (
+        report_df
+            .groupby(groups)
+            .agg(
+                F.sum('hours').alias('hours'), 
+                F.sum('search_counts.google').alias('google'),
+                F.sum('search_counts.bing').alias('bing'),
+                F.sum('search_counts.yahoo').alias('yahoo'),
+                F.sum('search_counts.other').alias('other')) 
+            .join(crash_agg_df, ['country', 'channel', 'os', 'date'])
```

Comment:
`crash_agg_df` doesn't appear to have a `date` column, does that impact this join (or the `groupby` that uses that column as well)?